stdlib/public/Concurrency/AsyncCompactMapSequence.swift

```diff
@@ -0,0 +1,80 @@
+////===----------------------------------------------------------------------===//
+////
+//// This source file is part of the Swift.org open source project
+////
+//// Copyright (c) 2021 Apple Inc. and the Swift project authors
+//// Licensed under Apache License v2.0 with Runtime Library Exception
+////
+//// See https://swift.org/LICENSE.txt for license information
+//// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+////
+////===----------------------------------------------------------------------===//
+
+import Swift
+
+extension AsyncSequence {
+  @inlinable
+  public __consuming func compactMap<ElementOfResult>(
+    _ transform: @escaping (Element) async -> ElementOfResult?
+  ) -> AsyncCompactMapSequence<Self, ElementOfResult> {
+    return AsyncCompactMapSequence(self, transform: transform)
+  }
+}
+
+@frozen
+public struct AsyncCompactMapSequence<Upstream: AsyncSequence, ElementOfResult> {
+  public let upstream: Upstream
+  public let transform: (Upstream.Element) async -> ElementOfResult?
+
+  @inlinable
+  public init(
+    _ upstream: Upstream, 
+    transform: @escaping (Upstream.Element) async -> ElementOfResult?
+  ) {
+    self.upstream = upstream
+    self.transform = transform
+  }
+}
+
+extension AsyncCompactMapSequence: AsyncSequence {
+  public typealias Element = ElementOfResult
+  public typealias AsyncIterator = Iterator
+
+  @frozen
+  public struct Iterator: AsyncIteratorProtocol {
+    public typealias Element = ElementOfResult
+
+    @usableFromInline
+    var upstreamIterator: Upstream.AsyncIterator
+
+    @usableFromInline
+    let transform: (Upstream.Element) async -> ElementOfResult?
+
+    @usableFromInline
+    init(
+      _ upstreamIterator: Upstream.AsyncIterator, 
+      transform: @escaping (Upstream.Element) async -> ElementOfResult?
+    ) {
+      self.upstreamIterator = upstreamIterator
+      self.transform = transform
+    }
+
+    @inlinable
+    public mutating func next() async rethrows -> ElementOfResult? {
+      while true {
```

Comment:
I noticed failable compact map will nil the closure on failure, which is also the end of the stream. Do we not want to throw away this closure on end of stream (nil) as well? Or, alternatively, not nil the closure on failure for the bailable version?

Reply:
it was decided to avoid niling out the closures and use a tracking state so that the closures always live as long as the stream.

Reply:
FWIW, when documenting this, let's go with something like "the closure may live as long as the iterator" rather than "always lives as long as the iterator".